When I run all the tests in `tests/rustdoc-gui` with command on my dev machine:
```console
./x.py test tests/rustdoc-gui --stage 1
```

There are some failed with errors, I verified it's not related to my new added test case:


```console
➜  rust git:(yukang-fix-sidebar-sort) ./x.py test tests/rustdoc-gui --stage 1
Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.11s
Building stage0 library artifacts (aarch64-apple-darwin)
    Finished `release` profile [optimized + debuginfo] target(s) in 0.04s
Building compiler artifacts (stage0 -> stage1, aarch64-apple-darwin)
    Finished `release` profile [optimized + debuginfo] target(s) in 0.44s
Creating a sysroot for stage1 compiler (use `rustup toolchain link 'name' build/host/stage1`)
Building stage1 library artifacts (aarch64-apple-darwin)
    Finished `release` profile [optimized + debuginfo] target(s) in 0.03s
Building stage0 tool rustdoc-gui-test (aarch64-apple-darwin)
    Finished `release` profile [optimized + debuginfo] target(s) in 0.19s
Building tool rustdoc (stage0 -> stage1, aarch64-apple-darwin)
    Finished `release` profile [optimized + debuginfo] target(s) in 0.16s
Testing rustdoc-gui (stage1 -> stage2, aarch64-apple-darwin)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.02s
   Generated /Users/yukang/rust/build/aarch64-apple-darwin/test/rustdoc-gui/doc/settings/index.html
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.05s
   Generated /Users/yukang/rust/build/aarch64-apple-darwin/test/rustdoc-gui/doc/lib2/index.html
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.01s
   Generated /Users/yukang/rust/build/aarch64-apple-darwin/test/rustdoc-gui/doc/link_to_definition/index.html
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.01s
   Generated /Users/yukang/rust/build/aarch64-apple-darwin/test/rustdoc-gui/doc/huge_logo/index.html
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.14s
   Generated /Users/yukang/rust/build/aarch64-apple-darwin/test/rustdoc-gui/doc/scrape_examples/index.html
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.01s
   Generated /Users/yukang/rust/build/aarch64-apple-darwin/test/rustdoc-gui/doc/proc_macro_test/index.html
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.01s
   Generated /Users/yukang/rust/build/aarch64-apple-darwin/test/rustdoc-gui/doc/staged_api/index.html
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.01s
   Generated /Users/yukang/rust/build/aarch64-apple-darwin/test/rustdoc-gui/doc/extend_css/index.html
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.01s
   Generated /Users/yukang/rust/build/aarch64-apple-darwin/test/rustdoc-gui/doc/theme_css/index.html
   Compiling test_docs v0.1.0 (/Users/yukang/rust/tests/rustdoc-gui/src/test_docs)
 Documenting test_docs v0.1.0 (/Users/yukang/rust/tests/rustdoc-gui/src/test_docs)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 1.18s
   Generated /Users/yukang/rust/build/aarch64-apple-darwin/test/rustdoc-gui/doc/test_docs/index.html
Running 136 rustdoc-gui (10 concurrently) ...
.......... (10/136)
.......... (20/136)
.......... (30/136)
.......... (40/136)
.......... (50/136)
.......... (60/136)
.......... (70/136)
.......... (80/136)
.......... (90/136)
.......... (100/136)
.......... (110/136)
.......... (120/136)
.......... (130/136)
FFFFFF     (136/136)


/Users/yukang/rust/tests/rustdoc-gui/item-info.goml item-info... FAILED
[ERROR] `tests/rustdoc-gui/item-info.goml` line 23: Condition `|first_line_y| != |second_line_y| && |first_line_y| == 718 && |second_line_y| == 741` (`718 != 742 && 718 == 718 && 742 == 741`) was evaluated as false: for command `assert: |first_line_y| != |second_line_y| && |first_line_y| == 718 && |second_line_y| == 741`

/Users/yukang/rust/tests/rustdoc-gui/search-result-display.goml search-result-display... FAILED
[WARNING] `tests/rustdoc-gui/search-result-display.goml` line 37: Delta is 0 for "x", maybe try to use `compare-elements-position` instead?

[ERROR] `tests/rustdoc-gui/search-result-display.goml` line 53: The following errors happened (for selector `#crate-search`): [expected `159px` for key `width`, found `158.5px`]: for command `assert-css: ("#crate-search", {"width": "159px"})`

/Users/yukang/rust/tests/rustdoc-gui/search-filter.goml search-filter... FAILED
[ERROR] `tests/rustdoc-gui/search-filter.goml` line 24: The following errors happened: [document property `URL` (`file:///Users/yukang/rust/build/aarch64-apple-darwin/test/rustdoc-gui/doc/test_docs/index.html?search=test`) doesn't contain `&filter-crate=` (for CONTAINS check)]: for command `assert-document-property: ({"URL": "&filter-crate="}, CONTAINS)`
[ERROR] `tests/rustdoc-gui/search-filter.goml` line 26: assert didn't fail: for command `assert-false: "#results .externcrate"`
[ERROR] `tests/rustdoc-gui/search-filter.goml` line 28: The following errors happened (for selector `#crate-search`): [expected `lib2` for property `"value"`, found `all crates`]: for command `assert-property: ("#crate-search", {"value": "lib2"})`
[ERROR] `tests/rustdoc-gui/search-filter.goml` line 37: assert didn't fail: for command `assert-false: "#results .externcrate"`
[ERROR] `tests/rustdoc-gui/search-filter.goml` line 38: The following errors happened (for selector `#crate-search`): [expected `lib2` for property `"value"`, found `all crates`]: for command `assert-property: ("#crate-search", {"value": "lib2"})`

/Users/yukang/rust/tests/rustdoc-gui/search-tab.goml search-tab... FAILED
[ERROR] `tests/rustdoc-gui/search-tab.goml` line 81: The following errors happened: [` (28) ` isn't equal to ` (27) `]: for command `assert-text: ("#search-tabs > button:nth-child(1) > .count", " (27) ")`

/Users/yukang/rust/tests/rustdoc-gui/sidebar-mobile.goml sidebar-mobile... FAILED
[ERROR] `tests/rustdoc-gui/sidebar-mobile.goml` line 52: The following errors happened: [different Y values: 45 (or 45) != 46]: for command `assert-position: ("#method\.must_use", {"y": 46})`

/Users/yukang/rust/tests/rustdoc-gui/sidebar-source-code-display.goml sidebar-source-code-display... FAILED
[ERROR] `tests/rustdoc-gui/sidebar-source-code-display.goml` line 145: The following errors happened: [expected `2578` for window property `pageYOffset`, found `2578.5`]: for command `assert-window-property: {"pageYOffset": |y_offset|}`
[ERROR] `tests/rustdoc-gui/sidebar-source-code-display.goml` line 152: The following errors happened: [expected `2578` for window property `pageYOffset`, found `2578.5`]: for command `assert-window-property: {"pageYOffset": |y_offset|}`

Error: ()
Build completed unsuccessfully in 0:01:00
```

Is it because of my own browser settings, I don't know.

--- a/src/librustdoc/html/render/sidebar.rs
+++ b/src/librustdoc/html/render/sidebar.rs
@@ -1,4 +1,5 @@
 use std::borrow::Cow;
+use std::cmp::Ordering;
 
 use rinja::Template;
 use rustc_data_structures::fx::FxHashSet;
@@ -12,6 +13,7 @@ use crate::clean;
 use crate::formats::Impl;
 use crate::formats::item_type::ItemType;
 use crate::html::markdown::{IdMap, MarkdownWithToc};
+use crate::html::render::print_item::compare_names;
 
 #[derive(Clone, Copy)]
 pub(crate) enum ModuleLike {
@@ -77,7 +79,7 @@ impl<'a> LinkBlock<'a> {
 }
 
 /// A link to an item. Content should not be escaped.
-#[derive(PartialOrd, Ord, PartialEq, Eq, Hash, Clone)]
+#[derive(Ord, PartialEq, Eq, Hash, Clone)]
 pub(crate) struct Link<'a> {
     /// The content for the anchor tag and title attr
     name: Cow<'a, str>,
@@ -87,6 +89,20 @@ pub(crate) struct Link<'a> {
     href: Cow<'a, str>,
     /// Nested list of links (used only in top-toc)
     children: Vec<Link<'a>>,
+}
+
+impl PartialOrd for Link<'_> {
+    fn partial_cmp(&self, other: &Link<'_>) -> Option<Ordering> {
+        match compare_names(&self.name, &other.name) {
+            Ordering::Equal => (),
+            result => return Some(result),
+        }
+        (&self.name_html, &self.href, &self.children).partial_cmp(&(
+            &other.name_html,
+            &other.href,
+            &other.children,
+        ))
+    }
 }
 
 impl<'a> Link<'a> {

--- a/tests/rustdoc-gui/sidebar-foreign-impl-sort.goml
+++ b/tests/rustdoc-gui/sidebar-foreign-impl-sort.goml
@@ -1,0 +1,15 @@
+// Checks sidebar resizing close the Settings popover
+go-to: "file://" + |DOC_PATH| + "/test_docs/SidebarSort/trait.Sort.html#foreign-impls"
+
+// Check that the sidebar contains the expected foreign implementations
+assert-text: (".sidebar-elems section ul > li:nth-child(1)", "&'a str")
+assert-text: (".sidebar-elems section ul > li:nth-child(2)", "AtomicBool")
+assert-text: (".sidebar-elems section ul > li:nth-child(3)", "AtomicU8")
+assert-text: (".sidebar-elems section ul > li:nth-child(4)", "AtomicU16")
+assert-text: (".sidebar-elems section ul > li:nth-child(5)", "AtomicU32")
+assert-text: (".sidebar-elems section ul > li:nth-child(6)", "Cell<u8>")
+assert-text: (".sidebar-elems section ul > li:nth-child(7)", "Cell<u16>")
+assert-text: (".sidebar-elems section ul > li:nth-child(8)", "u8")
+assert-text: (".sidebar-elems section ul > li:nth-child(9)", "u16")
+assert-text: (".sidebar-elems section ul > li:nth-child(10)", "u32")
+assert-text: (".sidebar-elems section ul > li:nth-child(11)", "usize")

--- a/tests/rustdoc-gui/src/test_docs/lib.rs
+++ b/tests/rustdoc-gui/src/test_docs/lib.rs
@@ -713,3 +713,21 @@ pub trait ItemsTrait {
     /// blablala
     fn bar();
 }
+
+pub mod SidebarSort {
+    use std::cell::Cell;
+    use std::sync::atomic::*;
+    pub trait Sort {}
+
+    impl Sort for u32 {}
+    impl Sort for u8 {}
+    impl Sort for u16 {}
+    impl Sort for usize {}
+    impl Sort for AtomicU32 {}
+    impl Sort for AtomicU16 {}
+    impl Sort for AtomicU8 {}
+    impl Sort for AtomicBool {}
+    impl Sort for Cell<u16> {}
+    impl Sort for Cell<u8> {}
+    impl<'a> Sort for &'a str {}
+}

